docs/markdown/Rewriter.md

```diff
@@ -0,0 +1,183 @@
+---
+short-description: Automatic modification of the build system files
+...
+
+# Meson file rewriter
+
+Since version 0.50.0, meson has the functionality to perform some basic
+modification on the `meson.build` files from the command line. The currently
+supported operations are:
+
+- For build targets:
+  - Add/Remove source files
+  - Add/Remove targets
+  - Modify a select set of kwargs
+  - Print some JSON information
+- For dependencies:
+  - Modify a select set of kwargs
+- For the project function:
+  - Modify a select set of kwargs
+  - Modify the default options list
+
+The rewriter has both, a normal command line interface and a "script mode". The
+normal CLI is mostly designed for everyday use. The "script mode", on the
+other hand, is meant to be used by external programs (IDEs, graphical
+frontends, etc.)
+
+## Using the rewriter
+
+All rewriter functions are accessed via `meson rewrite` (or the rewriter alias
+`meson rw`). The meson rewriter assumes that it is run inside the project root
+directory. If this isn't the case, use `--sourcedir` to specify the actual
+project source directory.
+
+### Adding and removing sources
+
+The most common operations will probably be the adding and removing of source
+files to a build target. This can be easily done with:
+
+```bash
+meson rewrite target <target name/id> {add/rm} [list of sources]
+```
+
+For instance, given the following example
+
+```meson
+src = ['main.cpp', 'fileA.cpp']
+
+exe1 = executable('testExe', src)
+```
+
+the source `fileB.cpp` can be added with:
+
+```bash
+meson rewrite target testExe add fileB.cpp
+```
+
+After executing this command, the new `meson.build` will look like this:
```

Comment:
Best practice is to have each file on it's own line. Currently this reformats the whole list to be one one line. I really dislike that.

I also think the rewriter should add new sources in their alphabetical position. This *really* reduces merge conflicts.

Reply:
The problem with putting the sources on a new line is that this would mean that __all__ arrays would have line breaks. This is due to the internal implementation of the rewriter.

Sorting the sources could be implemented, but this would sort __all__ of the target on the first invokation of the rewriter. Are you (and I guess also @jpakkane) OK with this?

Reply:
Our rewriting in the qtcreator plugin is much more restricted and opinionated, because it only sees lists that are one file per line and properly marked. It also just leaves all lines that look unexpected as is.

For file lists i'm ok with enforcing them to be sorted. But i don't think it's technically needed. The rewriter *could* also just look for the first position in the lists where `current item < new item < new item` holds and insert there (edge cases apply).

For one item per line in the lists maybe we can just save a bool if the array had more than one item per line and switch between using the current layout and a layout with one item per line and an fixed two spaces indentation?

Reply:
If we sort file lists, I would enforce sorting the entire list. I am not a fan of partially sorted lists. Having the rewriter preserve line breaks can be implemented, but this would take some effort (so no `0.50.0`). Also, the rewriter automatically does linebreaks when there are more than 5 (I think) elements in a list.

Reply:
For new projects i absolutely would go with a completely sorted list for IDE integration¹. Existing projects might not be to happy about PRs/commits that sort the whole list though.

¹ the mini syntax in the qtcreator plugin does completly listed only. Because it requires by in anyway that seemed like the right choice.

Reply:
Maybe we can add a rewriter configuration (like .clang-format) file in the future to disable sorting, set indentation, etc.

Reply:
Yes, something like that would be an escape hatch, in case we get too much push back against sorting in the future and we want to accommodate that.

Reply:
In that case, I will implement the file sorting in a new PR.

Reply:
Sorting PR is #5010